src/api/procedures/setTransferRestrictions.ts

```diff
@@ -18,88 +25,128 @@ import {
 }
-
 /**
  * @hidden
  */
-function transformRestrictions(
+function transformInput(
   restrictions:
     | CountTransferRestrictionInput[]
     | PercentageTransferRestrictionInput[]
     | ClaimCountTransferRestrictionInput[]
     | ClaimPercentageTransferRestrictionInput[],
   currentRestrictions: TransferCondition[],
+  currentExemptions: ExemptionRecords,
   type: TransferRestrictionType,
   context: Context
-): [PolymeshPrimitivesTransferComplianceTransferCondition[], (string | Identity)[]] {
-  const exemptions: (string | Identity)[] = [];
+): [PolymeshPrimitivesTransferComplianceTransferCondition[], ExemptionRecords, ExemptionRecords] {
+  const toSet: ExemptionRecords = {};
+  const toRemove: ExemptionRecords = {};
 
-  let someDifference = restrictions.length !== currentRestrictions.length;
+  let needDifferentConditions = restrictions.length !== currentRestrictions.length;
   const conditions: PolymeshPrimitivesTransferComplianceTransferCondition[] = [];
+  const givenExemptions: ExemptionRecords = {};
+
   restrictions.forEach(r => {
     let value: BigNumber | ClaimCountRestrictionValue | ClaimPercentageTransferRestriction;
+    let claimType: ClaimType | undefined;
     if ('count' in r) {
       value = r.count;
     } else if ('percentage' in r) {
       value = r.percentage;
     } else {
       value = r;
+      claimType = r.claim.type;
     }
 
     const condition = { type, value } as TransferRestriction;
 
     const compareConditions = (transferCondition: TransferCondition): boolean =>
       compareTransferRestrictionToInput(transferCondition, condition);
-    if (!someDifference) {
-      someDifference = ![...currentRestrictions].find(compareConditions);
+    if (!needDifferentConditions) {
+      needDifferentConditions = ![...currentRestrictions].find(compareConditions);
     }
     const rawCondition = transferRestrictionToPolymeshTransferCondition(condition, context);
+    conditions.push(rawCondition);
 
-    if (r.exemptedIdentities) {
-      r.exemptedIdentities.forEach(e => {
-        exemptions.push(e);
-      });
-    }
+    r.exemptedIdentities?.forEach(e => {
+      const key = claimType || 'None';
+      const id = typeof e === 'string' ? new Identity({ did: e }, context) : e;
+      addExemptionIfNotPresent(id, givenExemptions, key);
+    });
+  });
 
-    conditions.push(rawCondition);
+  Object.entries(givenExemptions).forEach(([cType, toAddIds]) => {
+    const currentIds = currentExemptions[cType as ClaimKey] || [];
+    toAddIds.forEach(id => {
+      addExemptionIfNotPresent(id, toSet, cType as ClaimKey, currentIds);
+    });
   });
 
-  if (!someDifference) {
+  Object.entries(currentExemptions).forEach(([cType, currentIds]) => {
+    const given = givenExemptions[cType as ClaimKey] || [];
+    currentIds.forEach(id => {
+      addExemptionIfNotPresent(id, toRemove, cType as ClaimKey, given);
+    });
+  });
+
+  const sizeOf = (obj: Record<string, unknown>): number => Object.keys(obj).length;
```

Comment:
`forEach` was adding a `| undefined` type vs `Object.entries(...).forEach` giving the type.

Reply:
If you do what I specified in https://github.com/PolymeshAssociation/polymesh-sdk/pull/807#discussion_r943732849, you can remove the `Partial` from the `ExemptionRecords` type, which will remove the `| undefined` from the `forEach`. This will mean that you will have to initialize your `ExemptionRecord` objects with empty arrays for each property and alter the logic slightly. It might seem a bit unnecessary but you'll get cleaner types